**LOADING THE DATA**

In [1]:
import pandas as pd
import numpy as np

In [2]:
pd.set_option('display.max_columns', 20)
df = pd.read_csv('clean_data.csv').drop(['ID'],axis=1)
df

Gender  Own_car  Own_property  Work_phone  Phone  Email  Unemployed  \
0          1        1             1           1      0      0           0   
1          1        1             1           0      0      0           0   
2          0        0             1           0      1      1           0   
3          0        0             1           0      0      0           1   
4          1        1             1           1      1      1           0   
...      ...      ...           ...         ...    ...    ...         ...   
9704       0        0             0           0      0      0           0   
9705       0        0             1           1      1      0           0   
9706       1        1             1           0      0      0           0   
9707       0        0             1           0      1      1           0   
9708       1        0             1           0      0      0           0   

      Num_children  Num_family  Account_length  Total_income        Age  \
0                0           2              15      427500.0  32.868574   
1                0           2              29      112500.0  58.793815   
2                0           1               4      270000.0  52.321403   
3                0           1              20      283500.0  61.504343   
4                0           2               5      270000.0  46.193967   
...            ...         ...             ...           ...        ...   
9704             0           2              20      180000.0  56.400884   
9705             0           2              19      112500.0  43.360233   
9706             0           2              21       90000.0  52.296762   
9707             0           2              32      157500.0  33.914454   
9708             0           1              13      112500.0  25.155890   

      Years_employed           Income_type                 Education_type  \
0          12.435574               Working               Higher education   
1           3.104787               Working  Secondary / secondary special   
2           8.353354  Commercial associate  Secondary / secondary special   
3           0.000000             Pensioner               Higher education   
4           2.105450               Working               Higher education   
...              ...                   ...                            ...   
9704        0.542106             Pensioner  Secondary / secondary special   
9705        7.375921  Commercial associate  Secondary / secondary special   
9706        4.711938               Working  Secondary / secondary special   
9707        3.627727             Pensioner               Higher education   
9708        3.266323               Working  Secondary / secondary special   

             Family_status         Housing_type Occupation_type  Target  
0           Civil marriage     Rented apartment           Other       1  
1                  Married    House / apartment  Security staff       0  
2     Single / not married    House / apartment     Sales staff       0  
3                Separated    House / apartment           Other       0  
4                  Married    House / apartment     Accountants       0  
...                    ...                  ...             ...     ...  
9704        Civil marriage  Municipal apartment        Laborers       1  
9705               Married    House / apartment           Other       1  
9706               Married    House / apartment           Other       1  
9707               Married    House / apartment  Medicine staff       1  
9708  Single / not married     Rented apartment        Laborers       1  

[9709 rows x 19 columns]

**DATA PREPROCESSING**

In [3]:
df.columns

Index(['Gender', 'Own_car', 'Own_property', 'Work_phone', 'Phone', 'Email',
       'Unemployed', 'Num_children', 'Num_family', 'Account_length',
       'Total_income', 'Age', 'Years_employed', 'Income_type',
       'Education_type', 'Family_status', 'Housing_type', 'Occupation_type',
       'Target'],
      dtype='object')

In [4]:
df = pd.get_dummies(df,columns=['Income_type',
       'Education_type', 'Family_status', 'Housing_type', 'Occupation_type']).replace({True:1,False:0})

In [5]:
df

Gender  Own_car  Own_property  Work_phone  Phone  Email  Unemployed  \
0          1        1             1           1      0      0           0   
1          1        1             1           0      0      0           0   
2          0        0             1           0      1      1           0   
3          0        0             1           0      0      0           1   
4          1        1             1           1      1      1           0   
...      ...      ...           ...         ...    ...    ...         ...   
9704       0        0             0           0      0      0           0   
9705       0        0             1           1      1      0           0   
9706       1        1             1           0      0      0           0   
9707       0        0             1           0      1      1           0   
9708       1        0             1           0      0      0           0   

      Num_children  Num_family  Account_length  ...  \
0                0           2              15  ...   
1                0           2              29  ...   
2                0           1               4  ...   
3                0           1              20  ...   
4                0           2               5  ...   
...            ...         ...             ...  ...   
9704             0           2              20  ...   
9705             0           2              19  ...   
9706             0           2              21  ...   
9707             0           2              32  ...   
9708             0           1              13  ...   

      Occupation_type_Low-skill Laborers  Occupation_type_Managers  \
0                                      0                         0   
1                                      0                         0   
2                                      0                         0   
3                                      0                         0   
4                                      0                         0   
...                                  ...                       ...   
9704                                   0                         0   
9705                                   0                         0   
9706                                   0                         0   
9707                                   0                         0   
9708                                   0                         0   

      Occupation_type_Medicine staff  Occupation_type_Other  \
0                                  0                      1   
1                                  0                      0   
2                                  0                      0   
3                                  0                      1   
4                                  0                      0   
...                              ...                    ...   
9704                               0                      0   
9705                               0                      1   
9706                               0                      1   
9707                               1                      0   
9708                               0                      0   

      Occupation_type_Private service staff  Occupation_type_Realty agents  \
0                                         0                              0   
1                                         0                              0   
2                                         0                              0   
3                                         0                              0   
4                                         0                              0   
...                                     ...                            ...   
9704                                      0                              0   
9705                                      0                              0   
9706                                      0                              0   
9707                                      

In [6]:
df.isna().sum()

Gender                                          0
Own_car                                         0
Own_property                                    0
Work_phone                                      0
Phone                                           0
Email                                           0
Unemployed                                      0
Num_children                                    0
Num_family                                      0
Account_length                                  0
Total_income                                    0
Age                                             0
Years_employed                                  0
Target                                          0
Income_type_Commercial associate                0
Income_type_Pensioner                           0
Income_type_State servant                       0
Income_type_Student                             0
Income_type_Working                             0
Education_type_Academic degree                  0


In [7]:
df.columns

Index(['Gender', 'Own_car', 'Own_property', 'Work_phone', 'Phone', 'Email',
       'Unemployed', 'Num_children', 'Num_family', 'Account_length',
       'Total_income', 'Age', 'Years_employed', 'Target',
       'Income_type_Commercial associate', 'Income_type_Pensioner',
       'Income_type_State servant', 'Income_type_Student',
       'Income_type_Working', 'Education_type_Academic degree',
       'Education_type_Higher education', 'Education_type_Incomplete higher',
       'Education_type_Lower secondary',
       'Education_type_Secondary / secondary special',
       'Family_status_Civil marriage', 'Family_status_Married',
       'Family_status_Separated', 'Family_status_Single / not married',
       'Family_status_Widow', 'Housing_type_Co-op apartment',
       'Housing_type_House / apartment', 'Housing_type_Municipal apartment',
       'Housing_type_Office apartment', 'Housing_type_Rented apartment',
       'Housing_type_With parents', 'Occupation_type_Accountants',
       'Occupatio

In [8]:
input_cols = ['Gender', 'Own_car', 'Own_property', 'Work_phone', 'Phone', 'Email',
       'Unemployed', 'Num_children', 'Num_family', 'Account_length',
       'Total_income', 'Age', 'Years_employed',
       'Income_type_Commercial associate', 'Income_type_Pensioner',
       'Income_type_State servant', 'Income_type_Student',
       'Income_type_Working', 'Education_type_Academic degree',
       'Education_type_Higher education', 'Education_type_Incomplete higher',
       'Education_type_Lower secondary',
       'Education_type_Secondary / secondary special',
       'Family_status_Civil marriage', 'Family_status_Married',
       'Family_status_Separated', 'Family_status_Single / not married',
       'Family_status_Widow', 'Housing_type_Co-op apartment',
       'Housing_type_House / apartment', 'Housing_type_Municipal apartment',
       'Housing_type_Office apartment', 'Housing_type_Rented apartment',
       'Housing_type_With parents', 'Occupation_type_Accountants',
       'Occupation_type_Cleaning staff', 'Occupation_type_Cooking staff',
       'Occupation_type_Core staff', 'Occupation_type_Drivers',
       'Occupation_type_HR staff', 'Occupation_type_High skill tech staff',
       'Occupation_type_IT staff', 'Occupation_type_Laborers',
       'Occupation_type_Low-skill Laborers', 'Occupation_type_Managers',
       'Occupation_type_Medicine staff', 'Occupation_type_Other',
       'Occupation_type_Private service staff',
       'Occupation_type_Realty agents', 'Occupation_type_Sales staff',
       'Occupation_type_Secretaries', 'Occupation_type_Security staff',
       'Occupation_type_Waiters/barmen staff']
target_col = ['Target']

In [9]:
X = df[input_cols]
y = df[target_col]

In [10]:
from sklearn.preprocessing import MinMaxScaler

In [11]:
scaler = MinMaxScaler()
scaler.fit(X)

MinMaxScaler()

In [12]:
X = scaler.transform(X)

In [13]:
from sklearn.model_selection import train_test_split

In [14]:
X_train,X_test,y_train,y_test = train_test_split(X,y,test_size=0.25,random_state=42)

In [15]:
X_train = pd.DataFrame(X_train,columns=input_cols)
X_test = pd.DataFrame(X_test,columns=input_cols)

**TRAINING THE MODEL**

In [16]:
from sklearn.ensemble import RandomForestClassifier

In [17]:
model = RandomForestClassifier(n_jobs=-1,random_state=42)

In [18]:
model.fit(X_train,y_train)

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


RandomForestClassifier(n_jobs=-1, random_state=42)

In [19]:
model.score(X_train,y_train)

1.0

In [20]:
model.score(X_test,y_test)

0.8554365733113674

**LET'S CREATE A FUNCTION TO TUNE THE PARAMETERS**

In [21]:
def params(**kwargs):
    
    model = RandomForestClassifier(n_jobs=-1,random_state=42,**kwargs)
    model.fit(X_train,y_train)
    return model.score(X_train,y_train), model.score(X_test,y_test)    

In [22]:
params(n_estimators=60)

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(0.9994506249141601, 0.8554365733113674)

In [23]:
params(max_depth=5)

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(0.8718582612278534, 0.8558484349258649)

In [24]:
params(class_weight={1:1,0:7})

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(1.0, 0.8550247116968699)

In [25]:
params(max_samples=0.1)

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(0.8718582612278534, 0.8558484349258649)

**FINAL MODEL**

In [26]:
params(max_samples=0.1,max_depth=5,n_estimators=60,class_weight={1:1,0:7})

/home/apooyadv/.local/lib/python3.10/site-packages/sklearn/base.py:1152: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples,), for example using ravel().
  return fit_method(estimator, *args, **kwargs)


(0.8718582612278534, 0.8558484349258649)

We can see that our model's accuracy is close to 86% which is pretty decent considering the random nature of Credit Risk.